### PR TITLE
Red 12507 db migration

### PR DESCRIPTION
**Pull request by RasVandoros**

### CODE REVIEW

File: lib/LIMS2/Model/Plugin/Miseq.pm [hunk 1]
```diff
@@ -2,18 +2,128 @@ package LIMS2::Model::Plugin::Miseq;
...
-
 use Moose::Role;
-
 use Hash::MoreUtils qw( slice slice_def );
 use Const::Fast;
 use Try::Tiny;
 use Log::Log4perl qw( :easy );
 use namespace::autoclean;
 use LIMS2::Model::Util::Miseq qw( miseq_well_processes );
+use Data::Dumper;
```
Comment (PeterFCKeen) [inline lib/LIMS2/Model/Plugin/Miseq.pm]: Imported but not utilised

File: lib/LIMS2/Model/Plugin/Miseq.pm [hunk 2]
```diff
@@ -90,28 +207,37 @@ sub update_miseq_well_experiment {
...
+    $class->{nhej_reads} = check_undef($validated_params->{nhej_reads}, $hash_well->{nhej_reads});
+    $class->{hdr_reads} = check_undef($validated_params->{hdr_reads}, $hash_well->{hdr_reads});
+    $class->{mixed_reads} = check_undef($validated_params->{mixed_reads}, $hash_well->{mixed_reads});
+    $well->update($class);
 
     return;
 }
 
+
+
```
Comment (PeterFCKeen) [inline lib/LIMS2/Model/Plugin/Miseq.pm]: unnecessary spacing

File: lib/LIMS2/Model/Plugin/Miseq.pm [hunk 3]
```diff
@@ -412,4 +571,18 @@ sub primer_param {
...
+    if (defined $updated){
+           return $updated;
+    }
+    else{
+        if(defined $existing){
+            return $existing;
+        }
+    }
+    return "0";
+}
```
Comment (PeterFCKeen) [inline lib/LIMS2/Model/Plugin/Miseq.pm]: Ternary operator would do this in a line
x ? x : y

File: lib/LIMS2/Model/Util/ImportCrispressoQC.pm [hunk 4]
```diff
@@ -0,0 +1,749 @@
...
+
+#SUBS GO HERE
+sub header_hash {
+    my ( $header, @expected_titles ) = @_;
+
+    my @titles = split( /\t/, lc $header );
+    my @intersection = get_intersection( [ \@titles, \@expected_titles ] );
+    my %head;
+    %head = map { lc $titles[$_] => $_ }
+        0 .. $#titles;    #creates a hash that has all the elements of the array as keys and their index as values
```
Comment (PeterFCKeen) [inline lib/LIMS2/Model/Util/ImportCrispressoQC.pm]: my %head = map { lc $titles[$_] => $_ } @titles

File: lib/LIMS2/Model/Util/ImportCrispressoQC.pm [hunk 5]
```diff
@@ -0,0 +1,749 @@
...
+}
+
+sub migrate_crispresso_subs {
+    my ( $model, $jobout, $data ) = @_;
+    my $job = get_crispr($jobout);
+    my $row = {
+            id  =>   $data->{miseq_well_experiment}->{id},
+            crispr  =>  $job ->{crispr},
+            date_stamp  =>  $job->{date}
+        };
```
Comment (PeterFCKeen) [inline lib/LIMS2/Model/Util/ImportCrispressoQC.pm]: Indention is off

File: lib/LIMS2/Model/Util/ImportCrispressoQC.pm [hunk 6]
```diff
@@ -0,0 +1,749 @@
...
+    my $limit = 10;
+    my $counter = 0;#counts the number of frequencies within one file
+    open( my $file_to_read, "<", "$frequency_paths" ) or die "Cannot open frequency file";
+    chomp(my @lines = <$file_to_read>);
+    close $file_to_read or die "Cannot close frequency file";
+
+    my $header = shift(@lines);    #grab the header line that holds the titles of the columns
+
+    my @expected_titles
+        = ( 'aligned_sequence', 'nhej', 'unmodified', 'hdr', 'n_deleted', 'n_inserted', 'n_mutated', '#reads' );
```
Comment (PeterFCKeen) [inline lib/LIMS2/Model/Util/ImportCrispressoQC.pm]: one line  - @expected_titles = qw( aligned_sequence nhej unmodified hdr n_deleted n_inserted n_mutated #reads );

File: lib/LIMS2/Model/Util/ImportCrispressoQC.pm [hunk 7]
```diff
@@ -0,0 +1,749 @@
...
+                        }
+                        catch {
+                            warn "Error creating entry";
+                            $model->schema->txn_rollback;
+                        };
+                    }
+                );
+            }
+            else {
+                last;
```
Comment (PeterFCKeen) [inline lib/LIMS2/Model/Util/ImportCrispressoQC.pm]: for loop with a limit of 10 or length, which ever comes first

File: lib/LIMS2/Model/Util/ImportCrispressoQC.pm [hunk 8]
```diff
@@ -0,0 +1,749 @@
...
+    my $counter = 0;#counts the number of frequencies within one file
+    open( my $file_to_read, "<", "$frequency_paths" ) or die "Cannot open frequency file";
+    chomp(my @lines = <$file_to_read>);
+    close $file_to_read or die "Cannot close frequency file";
+
+    my $header = shift(@lines);    #grab the header line that holds the titles of the columns
+
+    my @expected_titles
+        = ( 'aligned_sequence', 'nhej', 'unmodified', 'hdr', 'n_deleted', 'n_inserted', 'n_mutated', '#reads' );
+    my %head = header_hash( $header, @expected_titles );
```
Comment (PeterFCKeen) [inline lib/LIMS2/Model/Util/ImportCrispressoQC.pm]: Looks like youre trying to allocate a 'reference' to each value so you can call back the correct one later. In perl you can zip together arrays of the same length to create a hash: 
https://stackoverflow.com/questions/3715957/how-can-i-assign-two-arrays-to-a-hash-in-perl
I can't think of an example in our code base but here's a js example
`var headers = rows.shift().split(",");
...
var rowData = rows[rowCount].split(",");
var rowObj = headers.reduce((obj, key, index) => ({...obj, [key]: rowData[index]}), {});`

File: lib/LIMS2/WebApp/Controller/API/PointMutation.pm [hunk 9]
```diff
@@ -293,4 +244,170 @@ sub crispr_seq {
...
+    my ($c, $miseq_well_experiment_hash) = @_;
+
+    my $limit = $c->request->param('limit');
+    my $frequency_rs = $c->model('Golgi')->schema->resultset('MiseqAllelesFrequency')->search( { miseq_well_experiment_id => $miseq_well_experiment_hash->{id} });
+    my $cou = $frequency_rs->count;
+    my @lines;
+    push @lines ,'Aligned Sequence,NHEJ,Unmodified,HDR,Deleted,Inserted,Mutated,Reads,%Reads';
+    $limit = $cou if ($limit > $cou);
+    if ($limit > 0){
+            for my $i (1..$limit){
```
[the remark this answers is not in the document]
Comment (RasVandoros) [reply lib/LIMS2/WebApp/Controller/API/PointMutation.pm]: Should be fine now.

File: lib/LIMS2/WebApp/Controller/User/PointMutation.pm [hunk 10]
```diff
@@ -157,18 +163,54 @@ sub point_mutation_allele : Path('/user/point_mutation_allele') : Args(0) {
         0 => 96,
         1 => 384,
     };
+    my $var;
```
[the remark this answers is not in the document]
Comment (RasVandoros) [reply lib/LIMS2/WebApp/Controller/User/PointMutation.pm]: Changed var to exp_hash and temp to indel_freq

File: root/site/user/pointmutation/point_mutation_allele.tt [hunk 11]
```diff
@@ -287,11 +366,18 @@ $(document).ready(function() {
...
-    for (var expIndex = 0; expIndex < exps.length; expIndex++) {
-        getImages(exps[expIndex], '[% indel %]');
-        //getSummary(exps[expIndex]);
-    }
+for (var expIndex = 0; expIndex < exps.length; expIndex++) {
+    var indels = '[% indel_stats %]';
+    indels = jsonify(indels);
+    var selection = exps[expIndex];
+    JSON.stringify(selection);
+    var data = indels[selection];
```
[the remark this answers is not in the document]
Comment (RasVandoros) [reply root/site/user/pointmutation/point_mutation_allele.tt]: var indels = '[% indel_stats %]';
indels = jsonify(indels);

Are now outside of the loop.

File: root/site/user/pointmutation/point_mutation_allele.tt [hunk 12]
```diff
@@ -333,6 +426,150 @@ function toggleTable(table) {
     expTable.style.display = (expTable.style.display == "table") ? "none" : "table";
 }
 
+function mysortfunction(a,b) {
```
[the remark this answers is not in the document]
Comment (RasVandoros) [reply root/site/user/pointmutation/point_mutation_allele.tt]: Took out the whole function. It is not used anywhere apparently.

File: script/migration_script.pl [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+use Text::CSV;
+use LIMS2::Model::Util::Miseq qw( miseq_well_processes convert_index_to_well_name );
+use List::Compare::Functional qw( get_intersection );
+use LIMS2::Model::Util::ImportCrispressoQC qw( 
+                                                get_data_from_file 
+                                                migrate_quant_file 
+                                                update_miseq_exp 
+                                                migrate_images 
+                                                migrate_frequencies
+                                            );
```
[the remark this answers is not in the document]
Comment (RasVandoros) [reply script/migration_script.pl]: I am not entirely sure what is the correct indentation style, so I swapped to a one-liner.

**Pull request merged**